src/BaseControllerV2.ts

```diff
@@ -43,20 +51,63 @@ export type IsJsonable<T> =
       never;
 
 /**
- * Controller class that provides state management and subscriptions
+ * An function to derive state.
+ *
+ * This function will accept one piece of the controller state (one property),
+ * and will return some derivation of that state.
+ *
+ * @param value - A piece of controller state
+ * @returns Something derived from controller state
+ */
+export type StateDeriver<T> = (value: IsJsonable<T>) => IsJsonable<Json>;
+
+/**
+ * State metadata.
+ *
+ * This metadata describes which parts of state should be persisted, and how to
+ * get an anonymized representation of the state.
+ */
+export type StateMetadata<T> = {
+  [P in keyof T]: StatePropertyMetadata<T[P]>;
+};
+
+/**
+ * Metadata for a single state property
+ *
+ * @property persist - Indicates whether this property should be persisted
+ *   (`true` for persistent, `false` for transient), or is set to a function
+ *   that derives the persistent state from the state.
+ * @property anonymous - Indicates whether this property is already anonymous,
+ *   (`true` for anonymous, `false` if it has potential to be personally
+ *   identifiable), or is set to a function that returns an anonymized
+ *   representation of this state.
+ */
+export interface StatePropertyMetadata<T> {
+  persist: boolean | StateDeriver<T>;
+  anonymous: boolean | StateDeriver<T>;
+}
+
+type Json = null | boolean | number | string | Json[] | { [prop: string]: Json } | Partial<Record<never, never>>;
```

Comment:
I believe this type includes all valid JSON, but it also includes _invalid_ JSON. As soon as _any_ Record/object type is included, all sorts of unserializable things like `Date` objects and classes are include, for reasons that are beyond me.

So this is used as the basis of the return type, but the `isJsonable<>` type is still used to wrap it to guarantee the result is valid JSON.